README
App-Unix-RPasswd

App::Unix::RPasswd is an application to change passwords on UNIX and 
UNIX-like servers on a simple, fast (parallellized) and secure (SSH) way.
Retrievable 'random' passwords unique for each server can be generated and
automatically applied. Because the salt is secret, the passwords can only be 
regenerated by authorized personnel.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::Unix::RPasswd

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-Unix-RPasswd

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-Unix-RPasswd

    CPAN Ratings
        http://cpanratings.perl.org/d/App-Unix-RPasswd

    Search CPAN
        http://search.cpan.org/dist/App-Unix-RPasswd/


LICENSE AND COPYRIGHT

Copyright (C) 2011 Claudio Ramirez

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



Changes
Revision history for App-Unix-RPasswd

0.3    	20110316
		- First CPAN release on an unsuspecting world.
0.31   	20110318
		- POD fixes.
		- Migration to Module::Install for easier building.
0.32	20110320
		- Removal of namespace::autoclean dependency for easier building.
		- Removal of DateTime dependency for easier building.
		- Migration to Mouse for easier building and size.
0.33	20110321
		- Fix cornercase base == 0.
		- Fix cornercase servername eq ''.
0.34	20110322
		- Fix esthetics end repport.
		- Moose->Mouse refactoring (no fast cli response workarounds needed).
		- New tests
0.35	20110401
		- Fix parameter sessions bug.
0.40	20120112
		- Comply to Solaris 11 minalpha default
0.41	20121022
		- Bump Module::Install
0.5	20130111
		- Migration to Moo
		- Pretty print
0.51	20130111
		- Typo
0.52	20130530
		- Pass Perl 5.18.0 tests


Build.PL
use strict;
use warnings;
use inc::latest 'Module::Build';

my $builder = Module::Build->new(
    module_name       => 'App::Unix::RPasswd',
    license           => 'gpl',
    dist_author_from  => 'lib/App/Unix/RPasswd.pm',
    dist_version_from => 'lib/App/Unix/RPasswd.pm',
    build_requires    => {
        'Test::More'    => 0,
    },
    requires => {
        'perl'                  => '5.10.0',
        'Crypt::PasswdMD5'      => 0,
        'DateTime'              => 0,
        'Expect'                => 0,
        'File::Basename'        => 0,
        'Getopt::Long'          => 0,
        'List::MoreUtils'       => 0,
        'Moose'                 => 0,
        'namespace::autoclean'  => 0,
        'Parallel::ForkManager' => 0,
    },
    add_to_cleanup     => ['App-Unix-RPasswd-*'],
    create_makefile_pl => 'traditional',
    script_files       => { 'script/rpasswd' => 'bin/rpasswd', },
);

if ( !$builder->is_windowsish() ) {
    my $shebang = '#!' . $builder->perl();
    my @args = ( "-i", "-ple", 's@^#!.+@' . $shebang . '@ if $. == 1',
        'script/rpasswd' );
    $builder->do_system( $builder->perl(), @args ) or die($!);
}

$builder->add_build_element('script');
$builder->create_build_script();


lib/App/Unix/RPasswd.pm
package App::Unix::RPasswd;

use 5.010000;
use feature ':5.10';
use Moo;
use Parallel::ForkManager;
use App::Unix::RPasswd::Connection;
use App::Unix::RPasswd::SaltedPasswd;

our $VERSION = '0.52';
our $AUTHOR  = 'Claudio Ramirez <nxadm@cpan.org>';

has 'args' => (
    is       => 'ro',
    #isa      => 'HashRef',
    required => 1,
);

has '_connection_obj' => (
    is      => 'ro',
    does    => 'App::Unix::RPasswd::Connection',
    default => sub {
        my $self = shift;
        App::Unix::RPasswd::Connection->new(
            user     => $self->args->{user},
            ssh_args => $self->args->{ssh},
        );
    },
    lazy     => 1,
    init_arg => undef,
);

has '_saltedpasswd_obj' => (
    is      => 'ro',
    does    => 'App::Unix::RPasswd::SaltedPasswd',
    default => sub {
        my $self = shift;
        App::Unix::RPasswd::SaltedPasswd->new( salt => $self->args->{base} );
    },
    lazy     => 1,
    init_arg => undef,
);

sub ask_key {
    my ( $self, $key ) = @_;
    my $ckeys = $key;
    $ckeys =~ s/(\w)(.+)/\U$1\E$2s/;    # key -> Keys
    my @msg =
      ( "Please introduce the $key: ", "\nPlease re-introduce the $key: " );
    my $counter    = 0;
    my $first_time = 1;
    print $msg[$counter];
    my @input;
    system( '/bin/stty', '-echo' );

    while (<STDIN>) {
        chomp;
        $input[$counter] = $_;
        if ( $counter == 1 ) {
            if ( $input[0] eq $input[1] ) { last }
            else {
                say "\n$ckeys are not the same...";
                $counter = 0;
            }
        }
        else { $counter++; $first_time = 0; }

        print $msg[$counter] unless $first_time;
    }
    system '/bin/stty echo';
    say '';
    return $input[0];
}

sub pexec {
    my $self    = shift;
    my @servers = @_;
    state $run++;
    my @errors;

    # Setup forks
    my $pfm = Parallel::ForkManager->new( $self->args->{sessions} );

    # Retrieve status
    $pfm->run_on_finish(
        sub {
            my ( undef, undef, undef, undef, undef, $dataref ) = @_;
            if ( defined $dataref and defined $$dataref ) {
                push @errors, $$dataref;
            }
        }
    );

    for my $server (@servers) {
        if (@servers) {
            $pfm->start() and next;
            my $error;
            if ( defined $self->args->{base} ) {
                $self->args->{password} =
                  $self->_saltedpasswd_obj->generate(
                    $server . $self->args->{date} . $server );
            }

            if ( $self->args->{generate_only} ) {
                say $self->args->{password} . " ($server)";
            }
            else {
                my $msg =
                  ( $self->args->{debug} )
                  ? 'Running on ' 
                  . $server
                  . ' with password \''
                  . $self->args->{password} . '\'...'
                  : "Running on $server...";
                say $msg;
                my $status = eval {
                    local $SIG{ALRM} = sub {
                        say "Timeout exceeded for $server.";
                        die "Timeout exceeded\n";    # NB: \n required
                    };
                    alarm $self->args->{timeout};
                    return $self->_connection_obj->run(
                        $server,
                        $self->args->{password},
                        $self->args->{debug}
                    );
                    alarm 0;
                };
                if ( !$status ) { $error = $server; }
            }
            $pfm->finish( 0, \$error );
        }
    }
    $pfm->wait_all_children;
    say "\nRun $run done.\n";
    return ( scalar @errors and $run <= $self->args->{reruns} )
      ? $self->pexec(@errors)
      : @errors;    # recursive
}

1;
__END__

=pod

=head1 NAME

App::Unix::RPasswd - Change passwords on UNIX and UNIX-like servers on a simple, 
fast (in parallel) and secure (SSH) way.

=head1 VERSION

Version 0.52

=head1 SYNOPSIS

App::Unix::RPasswd is an application for changing passwords on UNIX and 
UNIX-like servers on a simple, fast (in parallel) and secure (SSH) way. 
A salt-based retrievable "random" password generator, tied to the supplied 
server names and date, is included. 
This generated passwords, unique for each server, can be generated and
automatically remotely applied. Because the salt is secret and the correct date
string is required, the password for an specific server can only be regenerated 
by authorized personnel.

Perl 5.10 or higher is required.

The program has two modes. The default mode connects to remote targets and 
changes the password (optional) of the specified user (mandatory) on the 
supplied servers (mandatory). Optional valid parameters for this mode are 
sessions, ssh_args, reruns, timeout and debug. The built-in salted password 
generator can be used to create unique 'random' passwords for each server on 
the fly. In this case date (optional) and base (mandatory) are valid parameters 
for this mode.

The "generate_only" mode is used to (re-) generate salted passwords. In this 
mode only date (optional), base (mandatory), sessions (optional) and one of more 
servers (mandatory) are valid parameters.

From a security point of view, it is strongly advised to supply '-' as the base
salt or password on the command line. The program will then ask interactively 
for the base salt or password. This program requires a ssh-key based remote 
root access.

	Usage:
		rpasswd -u <user> -p <password> <server(s)>
		rpasswd -g -b <base salt> -date <YYYYMMDD> <server(s)>

	Options:
		--generate_only|-g: (re-)generate the salted password.
		--user|-u:          remote user name.
		--password|-p:      new password for remote user.
		--base|-b:          base salt for encryption.
		--date|-d:          date in YYYYMMDD format*.
		--ssh_args|-a:      settings for the ssh client*.
		--reruns|-r:        reruns for failed targets*.
		--sessions|-s:      simultaneous sessions*.
		--timeout|-t:       session timeout*.
		--debug:            prints debug output*.
		--help|-h:          print this help screen.
		--version|-v:       prints the version number.

		*: optional

=head1 PARAMETERS

=head2 generate_only | g

This parameter enables the (re-)generation of salted passwords.

=head2 user | u

This parameter sets the remote user name that will receive a new password.

=head2 password | p

This parameter sets the new password for the remote user. When "-" is supplied 
as argument, the program asks interactively for the password.

=head2 base | b

This parameter sets the base salt for encryption. When "-" is supplied as 
argument, the program asks interactively for the base salt. The salt can be 
between 1 and 8 characters. Longer salts are truncated.

=head2 date | d

This optional parameters sets the date string in a YYYYMMDD format (defaults to 
today).

=head2 ssh_args | a

This optional parameter sets additional settings for the ssh client (man ssh).
If you dont (locally) run the program as root but have root access via 
ssh-keys you need to use --ssh_args "-l root". Quote the argument string.

=head2 reruns | r

This optional parameterre sets the reruns for failed targets (defaults to 0).

=head2 sessions | s

This optional parameter sets the simultaneous sessions (defaults to 5).

=head2 timeout | t

This optional parameter sets the session timeout in seconds (defaults to 20 
seconds). While OpenSSH has the ConnectTimeout (passed as --ssh_args "-OConnectTimeout=<value>")
that provides a similar funcionality, its for on Solaris, SunSSH, has not. This
is a generic implementation that work on both ssh families.

=head2 debug

This parameter prints debug output.

=head2 help | h

This parameter prints this help screen.

=head2 version | v

This parameter prints the version number.

=head1 BUGS

Please report any bugs or feature requests to C<bug-App-Unix-RPasswd at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=App-Unix-RPasswd>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

This distribution has been tested on GNU/Linux (Debian 6.0 and Ubuntu 10.10) 
running OpenSSH and Solaris 10 running SunSSH.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App-Unix-RPasswd


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-Unix-RPasswd>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/App-Unix-RPasswd>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/App-Unix-RPasswd>

=item * Search CPAN

L<http://search.cpan.org/dist/App-Unix-RPasswd/>

=back

=head1 ACKNOWLEDGEMENTS

The following non-core modules were used:

=over 2

=item * Crypt::PasswdMD5
L<http://search.cpan.org/dist/Crypt-PasswdMD5/>

=item * Expect
L<http://search.cpan.org/dist/Expect/>

=item * List::MoreUtils
L<http://search.cpan.org/dist/List-MoreUtils/>

=item * Moo
L<http://search.cpan.org/dist/Moo/>

=item * Parallel::ForkManager
L<http://search.cpan.org/dist/Parallel-ForkManager/>

=back

=head1 AUTHOR

Claudio Ramirez <nxadm@cpan.org>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Claudio Ramirez.
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut


lib/App/Unix/RPasswd/Connection.pm
package App::Unix::RPasswd::Connection;
# This is an internal module of App::Unix::RPasswd

use feature ':5.10';
use Moo;
use Expect;

our $VERSION = '0.52';
our $AUTHOR  = 'Claudio Ramirez <nxadm@cpan.org>';

has 'user' => (
    is       => 'ro',
    #isa      => 'Str',
    required => 1,
);

has 'ssh_args' => (
    is       => 'ro',
    #isa      => 'ArrayRef[Str]',
    required => 1,
);

sub run {
    my ( $self, $server, $new_pass, $debug ) = @_;
    my $success = 0;
    my $exp = Expect->new();
    $exp->raw_pty(1);
    $exp->log_stdout(0) if !$debug;
    $exp->spawn( $self->_construct_cmd($server) )
      or warn 'Cannot change the password of '
      . $self->user
      . "\@$server: $!\n";
    $exp->expect(
        "10",
        [
            qr/password:/i => sub {
                my $exp = shift;
                $exp->send( $new_pass . "\r" );
                exp_continue;
              }
        ]
    );
    $exp->soft_close();
    $success = ( $exp->exitstatus == 0 ) ? 1 : 0;    # shell -> perl status
    if ( $success == 1 ) {
        say "Password changed on $server.";
    }
    else {
        warn "Failed to change the password on $server.\n";
    }
    return $success;
}

sub _construct_cmd {
    my ( $self, $server ) = @_;
    my @command = (
        @{ $self->ssh_args },
        $server, '/usr/bin/passwd', '-r', 'files', $self->user
    );
    return @command;
}

1;


lib/App/Unix/RPasswd/SaltedPasswd.pm
package App::Unix::RPasswd::SaltedPasswd;
# This is an internal module of App::Unix::RPasswd

use feature ':5.10';
use Moo;
use Crypt::PasswdMD5 ('unix_md5_crypt');
use List::MoreUtils  ('zip');

our $VERSION = '0.52';
our $AUTHOR  = 'Claudio Ramirez <nxadm@cpan.org>';

has 'salt' => (
    is       => 'ro',
    #isa      => 'Str',
    required => 1,
);

has 'minalpha' => ( # Minimum number of alpha character required
    is       => 'ro',
    #isa      => 'Int',
    default => sub { 2 },   # 2 is the default in Solaris 11
    required => 0,
);

sub generate {
    my ($self, $base_password)   = @_;
    
    # Create an encoded string
    my $passwd = $self->_encode_string(
        unix_md5_crypt( $base_password, $self->salt ) );
    
    # If necessary convert it to respect the minalpha constraint
    $passwd = $self->_minalpha_conv($passwd);
    return $passwd;
}

sub _encode_string {
    my ( $self, $opasswd ) = @_;
    $opasswd =~ tr/ /./;
    $opasswd =~ s/\$//g;
    my @array1  = split( //, $opasswd );
    my @array2  = reverse @array1;
    my @array3  = zip( @array2, @array1 );
    my $npasswd = join( '', @array3 );
    my $offset  = ( length $npasswd ) / 2 + 3;
    my $passwd  = substr( $npasswd, $offset, 12 ); # The password is 12 chars long
    return reverse $passwd;
}

sub _minalpha_conv {
    my ( $self, $opasswd ) = @_;
    my $passwd;
    my $first8_chars = substr($opasswd, 0, 8);
    if ($first8_chars !~ /[0-9]/) {
        my $ascii_value = ord(substr($first8_chars, 0, 1));
        my $sum;
        do {
            $sum = 0;
            my @digits = split(//,$ascii_value);
            for my $d (@digits) { $sum += $d; }
            $ascii_value = $sum;
        } while (length $sum != 1);
        $passwd = $sum . substr($opasswd, 1);
    }
    else { $passwd = $opasswd; }
    return $passwd;
   }

1;

# Additional properties of generated passwords
#MINDIFF=3      Minimum differences required between an old and a new password => OK (statically)
#MINALPHA=2     Minimum number of alpha character required => Done

# MAYBE TODO: make passwords suitable for new default configurations (e.g. Solaris 11)
#MINNONALPHA=1  Minimum number of non-alpha (including numeric and special) required
#MINUPPER=0     Minimum number of upper case letters tequired
#MINLOWER=0     Minimum number of lower case letters required
#MAXREPEATS=0   Maximum number of allowable consecutive repeating characters
#MINSPECIAL=0   Minimum number of special (non-alpha and non-digit) characters required
#MINDIGIT=0     Minimum number of digits required
#WHITESPACE=YES Determine if white space characters are allowed in passwords


lib/App/Unix/RPasswd/UI/Cli.pm
package App::Unix::RPasswd::UI::Cli;
# This is an internal module of App::Unix::RPasswd

use feature ':5.10';
use Moo;
use List::MoreUtils ('uniq');
use POSIX qw/strftime/;
use Term::ReadKey;

our $VERSION = '0.52';
our $AUTHOR  = 'Claudio Ramirez <nxadm@cpan.org>';

has 'args' => (
    is       => 'ro',
#    isa      => 'HashRef',
    required => 1,
);

has 'defaults' => (
    is       => 'ro',
#    isa      => 'HashRef',
    required => 1,
);

has 'messages' => (
    is       => 'rw',
#    isa      => 'ArrayRef',
    default  => sub { [] },
    reader   => 'get_messages',
    lazy     => 1,
    init_arg => undef,
);

has '_gen_only_options' => (
    is       => 'ro',
#    isa      => 'ArrayRef',
    default  => sub { [ 'base', 'date', 'sessions', 'generate_only', 'servers' ] },
    init_arg => undef,
);

sub check_params {

    # No params
    # Return success (1) or failure (O)
    my ( $self, $servers_ref ) = @_;
    my $status = 1;

    if ( !scalar @{$servers_ref} > 0 ) {
        $status = 0;
        push @{ $self->messages }, 'You need at least one server.';
    }
    return $status if !$status;

    # gen_only mode
    if ( $self->args->{generate_only} ) {
        $status = $self->_check_gen_only;
    }
    else {
        $status = $self->_check_main_mode;
    }
    return $status if !$status;

    # all modes
    $self->args->{ssh} = $self->defaults->{ssh};
    if ( defined $self->args->{ssh_args} ) {
        for my $arg ( split( /\s/, $self->args->{ssh_args} ) )
        {    # string to array
            push @{ $self->args->{ssh} }, $arg;
        }
    }

    if ( defined $self->args->{base} ) {    # Salts are valid for both modes
        if ( $self->args->{date} and $self->args->{date} !~ /^\d{8}$/ ) {
            $status = 0;
            push @{ $self->messages },
              'Supply parameter date in a YYYYMMDD format (e.g. 20101123).';
        }
        elsif ( !defined $self->args->{date} ) {
            $self->args->{date} = strftime "%Y%m%d", localtime;
        }
        if ( $self->args->{base} eq '-' ) {
            $self->args->{base} = $self->_ask_key('base salt');
        }
        
        while ( $self->args->{base} eq '' ) {
            say 'Base salt can not be empty.';
            $self->args->{base} = $self->_ask_key('base salt');
        }

    }
# This is an internal module of App::Unix::RPasswd
    return $status;
}

sub term_line {
    my $self = shift;
    my ($wchar) = GetTerminalSize();
    return "_" x $wchar . "\n";
}

sub show_help {
    my ( $self, $version_bool ) = @_;
    require File::Basename;
    my $program = File::Basename::basename($0);
    say $program . ', version ' . $VERSION . '.';
    return if $version_bool;
    my $reruns = $self->defaults->{runs} + 1;
    say <<"EOL";

Change passwords on UNIX and UNIX-like servers on a simple, fast (in parallel)
and secure (SSH) way. A salt-based retrievable "random" password generator, 
tied to the supplied server names and date, is included.

Usage:
\t$program -u <user> -p <password> <server(s)>
\t$program -g -b <base salt> -date <YYYYMMDD> <server(s)>

Options:
\t--generate_only|-g:\t(re-)generate the salted password.
\t--user|-u:\t\tremote user name.
\t--password|-p:\t\tnew password for remote user.
\t--base|-b:\t\tbase salt for encryption.
\t--date|-d:\t\tdate in YYYYMMDD format (defaults to today)*.
\t--ssh_args|-a:\t\tsettings for the ssh client (man ssh)*.
\t--reruns|-r:\t\treruns for failed targets (defaults to 0)*.
\t--sessions|-s:\t\tsimultaneous sessions (defaults to 5)*.
\t--timeout|-t:\t\tsession timeout (defaults to 20 seconds)*.
\t--debug:\t\tprints debug output*.
\t--help|-h:\t\tprints this help screen.
\t--version|-v:\t\tprints the version number.

\t*: optional

The program has two modes. The default mode connects to remote targets and 
changes the password (optional) of the specified user (mandatory) on the 
supplied servers (mandatory). Optional valid parameters for this mode are 
sessions, ssh_args ("-l root" if you don't the application as root), reruns, 
timeout and debug. The built-in salted password generator can be used to 
create unique 'random' passwords for each server on the fly. In this case 
date (optional) and base (mandatory) are valid parameters for this mode.

The "generate_only" mode is used to (re-) generate salted passwords. In this 
mode only date (optional), base (mandatory), sessions (optional) and one of 
more servers (mandatory) are valid parameters.

From a security point of view, it is strongly advised to supply '-' as the base
salt or password on the command line. The program will then ask interactively 
for the base salt or password. 

$AUTHOR, http://search.cpan.org/dist/App-Unix-RPasswd 
EOL
}

sub _ask_key {
    my ( $self, $key ) = @_;
    my $ckeys = $key;
    $ckeys =~ s/(\w)(.+)/\U$1\E$2s/;    # key -> Keys
    my @msg =
      ( "Please introduce the $key: ", "\nPlease re-introduce the $key: " );
    my $counter    = 0;
    my $first_time = 1;
    print $msg[$counter];
    my @input;
    system( '/bin/stty', '-echo' );

    while (<STDIN>) {
        chomp;
        $input[$counter] = $_;
        if ( $counter == 1 ) {
            if ( $input[0] eq $input[1] ) { last }
            else {
                say "\n$ckeys are not the same...";
                $counter = 0;
            }
        }
        else { $counter++; $first_time = 0; }

        print $msg[$counter] unless $first_time;
    }
    system '/bin/stty echo';
    say '';
    return $input[0];
}

sub _check_gen_only {
    my $self   = shift;
    my $status = 1;
    my @gen_options_provided =
      #grep { !( $_ ~~ @{ $self->_gen_only_options } ) }
      grep { !( /^$_$/, @{ $self->_gen_only_options } ) }
      keys %{ $self->args };
    for my $key (@gen_options_provided) {
        if ( defined $self->args->{$key} ) {
            $status = 0;
            push @{ $self->messages },
              "Parameter $key is invalid in this mode.";
        }
    }
    if ( !defined $self->args->{base} ) {
        $status = 0;
        push @{ $self->messages }, 'Parameter base is required in this mode.';
    }
    if ($status) {    # Reruns have no sense in this mode
        $self->args->{reruns} = 0;
    }
    return $status;
}

sub _check_main_mode {
    my $self   = shift;
    my $status = 1;

    if ( !defined $self->args->{user} ) {
        $status = 0;
        push @{ $self->messages }, 'Parameter user is mandatory.';
    }

    if (    !defined $self->args->{password}
        and !defined $self->args->{base} )
    {
        $status = 0;
        push @{ $self->messages }, 'You need to specify password or base.';
    }
    elsif ( defined $self->args->{password}
        and defined $self->args->{base} )
    {
        $status = 0;
        push @{ $self->messages },
          'You need to specify password or base, not both.';
    }
    return $status if !$status;

    if ( defined $self->args->{password} ) {
        if ( $self->args->{date} ) {
            $status = 0;
            push @{ $self->messages },
              'Date is only valid in combination with base.';
        }
        elsif ( $status == 1 and $self->args->{password} eq '-' ) {
            $self->args->{password} = $self->_ask_key('password');
        }
    }

    if ( !defined $self->args->{timeout} ) {
        $self->args->{timeout} = $self->defaults->{timeout};
    }

    if ( defined $self->args->{reruns} ) {
        if ( $self->args->{reruns} > 98 ) {
            $status = 0;
            push @{ $self->messages },
              'Less than 99 retries allowed. Let\'s be raisonable.';
        }
    }
    else { $self->args->{reruns} = 0 }
    return $status;
}

1;
